File: src/main/services/terminal/strategies/iterm2.ts
import { exec } from 'child_process'
import { promisify } from 'util'
import type { TerminalStrategy, EnvObject } from './base'

const execAsync = promisify(exec)

export class ITerm2Strategy implements TerminalStrategy {
  name = 'iTerm2'

  async isAvailable(): Promise<boolean> {
    try {
      // Check if iTerm2 is installed
      const { stdout } = await execAsync('mdfind "kMDItemCFBundleIdentifier == com.googlecode.iterm2"')
      return stdout.trim().length > 0
    } catch {
      return false
    }
  }

  async launch(path: string, env: EnvObject, cliCommand?: string): Promise<void> {
    // Escape path for shell
    const escapedPath = path.replace(/'/g, "'\\''")

    // Build inline environment variables (KEY="value" KEY2="value2" command)
    const envInline = Object.entries(env)
      .map(([key, value]) => {
        const escapedValue = value.replace(/"/g, '\\"')
        return `${key}="${escapedValue}"`
      })
      .join(' ')

    // Build the full command: cd to path, then run CLI with inline env vars
    let fullCommand = `cd '${escapedPath}' && clear`
    if (cliCommand) {
      fullCommand += ` && ${envInline} ${cliCommand}`
    }

    // Escape for AppleScript string
    const escapedCommand = fullCommand.replace(/\\/g, '\\\\').replace(/"/g, '\\"')

    const script = `
      tell application "iTerm2"
        activate
        set newWindow to (create window with default profile)
        tell current session of newWindow
          write text "${escapedCommand}"
        end tell
      end tell
    `

    await execAsync(`osascript -e '${script.replace(/'/g, "'\"'\"'")}'`)
  }
}

File: src-tauri/src/terminal/iterm2.rs
//! iTerm2 终端启动策略
//!
//! v3.2.0: 改用 wrapper 脚本方式。不再内联 env/token/CLI 命令到 AppleScript 中。
//! 终端只负责用 iTerm2 AppleScript 打开并执行 wrapper 脚本。

use super::{EnvObject, TerminalStrategy};
use std::process::Command;

pub struct ITerm2Strategy;

fn shell_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\"'\"'"))
}

impl TerminalStrategy for ITerm2Strategy {
    fn name(&self) -> &str {
        "iTerm2"
    }

    fn is_available(&self) -> bool {
        #[cfg(target_os = "macos")]
        {
            Command::new("mdfind")
                .args(["kMDItemCFBundleIdentifier == com.googlecode.iterm2"])
                .output()
                .map(|o| !o.stdout.is_empty())
                .unwrap_or(false)
        }
        #[cfg(not(target_os = "macos"))]
        {
            false
        }
    }

    fn launch(
        &self,
        working_dir: &str,
        _env: &EnvObject,
        wrapper_path: &str,
        _instance_label: Option<&str>,
    ) -> Result<(), String> {
        // v3.2.0: 直接执行 wrapper 脚本,不内联 env/command
        let command = format!(
            "cd {} && {}",
            shell_quote(working_dir),
            shell_quote(wrapper_path)
        );
        let script = format!(
            r#"tell application "iTerm2"
    activate
    set newWindow to (create window with default profile)
    tell current session of newWindow
        write text "{}"
    end tell
end tell"#,
            command.replace('"', "\\\"")
        );

        let output = Command::new("osascript")
            .args(["-e", &script])
            .output()
            .map_err(|e| format!("Failed to run osascript: {}", e))?;

        if output.status.success() {
            Ok(())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(format!("osascript failed: {}", stderr))
        }
    }
}
